From 8512269fa476bf6f68eead1712a0050c9615d365 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Mon, 15 Sep 2014 17:21:05 -0400
Subject: [redhat] cross: Updated cross compile information as it is now implemented

Message-id: <1410801665-5501-1-git-send-email-prarit@redhat.com>
Patchwork-id: 94800
O-Subject: [RHEL7.1 PATCH INTERNAL] Updated cross compile information as it is now implemented. [v4]
Bugzilla: 1117819
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Johnny Bieren <jbieren@redhat.com>

redhat/docs/RHELPatchRules was changed to reflect that make rh-cross-all-builds is now implemented in RHEL-7.1.
Also, redhat/Makefile.cross was edited to have variable EPEL_RELEASE.  Before the change, the following error occurred when running make rh-cross-all-builds:

rpm -ivh http://download.bos.redhat.com/pub/epel/beta/7/x86_64/epel-release-7-0.1.noarch.rpm
Retrieving http://download.bos.redhat.com/pub/epel/beta/7/x86_64/epel-release-7-0.1.noarch.rpm
Creating linux-3.10.0-133.el7.test.tar.xz...
curl: (22) The requested URL returned error: 404 Not Found
error: skipping http://download.bos.redhat.com/pub/epel/beta/7/x86_64/epel-release-7-0.1.noarch.rpm - transfer failed

In order to fix the error, the 7.0.1 in epel-release-7-0.1 needed to be variable to account for the latest updates to epel-release.

[v3]: text change recommended by jbenc
[v4]: update and change script name to download_cross.sh, dzickus

Cc: Don Zickus <dzickus@redhat.com>
Cc: Tony Camuso <tcamuso@redhat.com>
Cc: Stefan Assmann <sassmann@redhat.com>
Cc: Jes Sorensen <Jes.Sorensen@redhat.com>
Cc: Mike Snitzer <snitzer@redhat.com>
Cc: Jiri Benc <jbenc@redhat.com>
Cc: Daniel Borkmann <dborkman@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/redhat/Makefile.cross b/redhat/Makefile.cross
index 0cf56a2..d99a6b7 100644
--- a/redhat/Makefile.cross
+++ b/redhat/Makefile.cross
@@ -1,6 +1,3 @@
-EPEL_RELEASE=epel-release-7-0.1.noarch
-CROSS_EPEL_RPM = http://download.bos.redhat.com/pub/epel/beta/7/x86_64/$(EPEL_RELEASE).rpm
-
 CROSS_RPMFLAGS = $(RPMBUILD) --define "_sourcedir $(SOURCES)" --define "_builddir $(RPM)/BUILD" --define "_srcrpmdir $(RPM)/SRPMS" --define "_rpmdir $(RPM)/RPMS" --define "_specdir $(RPM)/SPECS" --define "dist $(DIST)"
 
 CROSS_PACKAGE_LIST = binutils-powerpc64-linux-gnu binutils-s390x-linux-gnu cross-binutils-common cross-gcc-common gcc-powerpc64-linux-gnu gcc-s390x-linux-gnu glibc-static
@@ -11,10 +8,7 @@ rh-cross-download:
   exit 1; \
  fi;
 ifeq ($(shell whoami), root)
-ifneq ($(shell rpm -q epel-release), $(EPEL_RELEASE))
- rpm -ivh $(CROSS_EPEL_RPM)
-endif
- @yum -y install $(CROSS_PACKAGE_LIST)
+ @$(REDHAT)/scripts/download_cross.sh $(CROSS_PACKAGE_LIST)
 else
  @echo "Checking for RHEL7 cross compile packages.  If this fails, run \"make rh-cross-download\" as root."
  @rpm -q $(CROSS_PACKAGE_LIST)
diff --git a/redhat/docs/RHELPatchRules b/redhat/docs/RHELPatchRules
index 95dc958..dacdf75 100644
--- a/redhat/docs/RHELPatchRules
+++ b/redhat/docs/RHELPatchRules
@@ -582,7 +582,25 @@ and submaintainers when dealing with RHEL.
 
 10. How do I cross compile a patch?  What is required?
 
-Cross compilation for RHEL7 is currently not implemented.
+In order to cross compile patches it is recommended that the latest nightly
+RHEL7 OS is installed.  Users can checkout a system from beaker or install a
+local system manually.  A user must also setup the "optional" repository for
+their RHEL7 install (note, that this is done automatically in a beaker
+install).
+
+After installing the system, a user must install the latest RHEL7 git tree,
+and apply their patches to this git tree.  After doing so, a user must type
+the following command as _root_:
+
+make rh-cross-download
+
+This will automatically install all dependencies for cross compiling.
+
+After this step is complete, a user can execute
+
+make rh-cross-all-builds
+
+which will build i686, s390x, and ppc64.
 
 VIII.  Useful Links
 -------------------
diff --git a/redhat/scripts/download_cross.sh b/redhat/scripts/download_cross.sh
new file mode 100755
index 0000000..81a7a6e
--- /dev/null
+++ b/redhat/scripts/download_cross.sh
@@ -0,0 +1,20 @@
+#!/bin/bash
+# Download & install the latest EPEL7 repo information
+# Argument(s) to this script is a list of rpm names to install
+
+# if everything is installed then exit successfully
+rpm -q $@ && exit 0
+
+# install epel-release if necessary
+rpm -q epel-release >& /dev/null
+if [ $? -ne 0 ]; then
+ wget -nd -r -l1 --no-parent -A "epel-release*.rpm" http://download.bos.redhat.com/pub/epel/beta/7/x86_64
+ rpm -ivh epel-release*.rpm
+ # clean up
+ rm -f epel-release*.rpm
+fi
+
+# install list of rpms for cross compile
+yum -y install $@
+
+exit 0
-- 
1.7.1